tools/codegen/api/cpp.py

```diff
@@ -36,90 +36,96 @@ def name(func: FunctionSchema, *, faithful_name_for_out_overloads: bool = False)
 # Translation of "value types" in JIT schema to C++ API type.  Value
 # types look the same no matter if they are argument types or return
 # types.  Returns None if the type in question is not a value type.
-def valuetype_type(t: Type) -> Optional[str]:
+def valuetype_type(t: Type, *, binds: ArgName) -> Optional[CType]:
     if isinstance(t, BaseType):
         if t.name == BaseTy.Tensor:
             return None
         elif t.name == BaseTy.int:
-            return 'int64_t'
+            return BaseCType('int64_t', binds)
         elif t.name == BaseTy.float:
-            return 'double'
+            return BaseCType('double', binds)
         elif t.name == BaseTy.str:
-            return 'std::string'
+            return BaseCType('std::string', binds)
         elif t.name in [BaseTy.bool, BaseTy.QScheme, BaseTy.Scalar,
                         BaseTy.ScalarType, BaseTy.Generator, BaseTy.Storage,
                         BaseTy.Layout, BaseTy.Device, BaseTy.MemoryFormat,
                         BaseTy.Dimname, BaseTy.Stream, BaseTy.ConstQuantizerPtr]:
             # These C++ names line up with their schema names
-            return t.name.name
+            return BaseCType(t.name.name, binds)
         else:
             raise AssertionError(f"unsupported type: {t}")
     elif isinstance(t, OptionalType):
-        elem = valuetype_type(t.elem)
+        elem = valuetype_type(t.elem, binds=binds)
         if elem is None:
             return None
-        return f"c10::optional<{elem}>"
+        return OptionalCType(elem)
     elif isinstance(t, ListType):
         if str(t.elem) == 'bool':
             assert t.size is not None
-            return f"std::array<bool,{t.size}>"
+            return BaseCType(f"std::array<bool,{t.size}>", binds)
         else:
             return None
     else:
         raise AssertionError(f"unrecognized type {repr(t)}")
 
 # Translation of types occuring in JIT arguments to a C++ argument type.
-def argumenttype_type(t: Type, *, mutable: bool) -> str:
+def argumenttype_type(t: Type, *, mutable: bool, binds: ArgName) -> CType:
```

Comment:
>An alternative would have been to construct CTypes without denotations and then fill them in post-facto, but I didn't do it this way. One downside is there are some places where I need a CType without denotation, so I fill these in with `__placeholder__` whenever this happens

How about returning a structured object that represents only the type without semantic information (e.g. `RawCType`). And `CType` could just be a simple data class:
```python
RawCType = Union[BaseRawCType, OptionalRawCType, ConstRefRawCType, MutRefRawCType]

@dataclass(frozen=True)
class CType:
    type: RawCppType
    name: ArgName
```

`argumenttype_type` can return a `RawCType` so it doesn't need to pass the `binds` argument everywhere. You also never need to pass `__placeholder__`. i.e. instead of 
```python
semantic_type = argument_type(t, binds=name)
str_type = argument_type(t, binds='__placeholder__').cpp_type()
```
you would have
```python
semantic_type = CType(type=argument_type(t), name=name)
string_type = argument_type(t).str()
```

Personally, I probably also rename `CType` to something else like `SemType` or `CArgument` so that `RawCType` could just be `CType`/`CppType`.

Reply:
The trouble with this is that it makes some operations on arguments require more unwrapping/wrapping. Imagine that you have `CType(ConstRefRawCType(BaseRawCType('Tensor')), 'other')`. A semantics preserving transformation would be to remove `ConstRefRawCType`, giving `CType(BaseRawCType('Tensor'), 'other')`. But I can't just say `orig_ctype.elem` as I did before; here I have to do `CType(orig_ctype.type.elem, orig_ctype.name)`.

Reply:
Also, seriously considered using SemType, but it's more characters XD XD XD

Reply:
If I'm reading this right, the actual `SemType` unwrapping is only done in one place (line 100 of transform.py). All other cases actually only care about the (non-semantic) C++ type information.

Either way, If it does come up a lot it could be made terser with some short helpers:
```python
def unwrap(self: SemType) -> SemType:
    return SemType(self.type.elem, self.name)

def wrap(self: SemType, kls: Callable) -> SemType:
    return SemType(kls(self.type), self.name)
```

Reply:
> If I'm reading this right, the actual SemType unwrapping is only done in one place (line 100 of transform.py). All other cases actually only care about the (non-semantic) type information.

That may very well be true how the code is today. There's some deeper aesthetic reason why I really want the ArgName on the leaves is because it also generalizes correctly when you have tuples: `PairCType(BaseRawCType('Tensor', 'one'), BaseRawCType('Tensor', 'two'))`. You have to do a parallel tree for annotations to represent this correctly. By the way, this is exactly how alias annotations were added to JIT schema type system (out-of-line) and I hate it haha.

By the way, the helper functions aren't well typed. You don't statically know that SemType's self.type is a class with an elem object. (And wrap is just ick!)